common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/model/immutable/ImmutableModelGenerator.java

```diff
@@ -0,0 +1,227 @@
+import javax.annotation.Nullable;
+import javax.lang.model.element.Modifier;
+import javax.tools.JavaFileObject;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static io.vrap.rmf.codegen.common.generator.util.CodeGeneratorUtil.*;
+
+public class ImmutableModelGenerator extends CodeGenerator {
+
+    protected final Logger LOGGER = LoggerFactory.getLogger(getClass());
+
+    public ImmutableModelGenerator(GeneratorConfig generatorConfig, Api api) {
+        super(generatorConfig, api);
+    }
+
+    private static String getMethodName(final Property property) {
+        final String javaPropertyName = Arrays.stream(property.getName().split("_"))
+                .map(StringUtils::capitalize)
+                .collect(Collectors.joining());
+        if (property.getType() instanceof BooleanType) {
+            return "is" + javaPropertyName;
+        }
+        return "get" + javaPropertyName;
+    }
+
+    public List<Path> getPaths(List<JavaFileObject> javaFiles, Path outputFolder) {
+        return javaFiles.stream().map(javaFile -> Paths.get(outputFolder.toString(), javaFile.getName())).collect(Collectors.toList());
+    }
+
+    protected Stream<AnnotationSpec> getDiscriminatorAnnotations(final ObjectType object) {
+        if (StringUtils.isEmpty(object.getDiscriminator()) || object.getSubTypes().isEmpty()) {
+            return Stream.empty();
+        }
+        final AnnotationSpec jsonTypeInfoAnnotation = AnnotationSpec.builder(JsonTypeInfo.class)
+                .addMember("use", "$T.NAME", JsonTypeInfo.Id.class)
+                .addMember("include", "$T.PROPERTY", JsonTypeInfo.As.class)
+                .addMember("property", "$S", object.getDiscriminator())
+                .addMember("visible", "true")
+                .build();
+
+        CodeBlock.Builder annotationBodyBuilder = CodeBlock.builder();
+        List<ObjectType> children = getSubtypes(object).collect(Collectors.toList());
+        if (!children.isEmpty()) {
+            for (int i = 0; i < children.size(); i++) {
+                annotationBodyBuilder.add(
+                        (i == 0 ? "{" : ",") +
+                                "\n@$T(value = $T.class, name = $S)"
+                                + (i == children.size() - 1 ? "\n}" : "")
+                        , JsonSubTypes.Type.class, getClassName(getPackagePrefix(), children.get(i)), children.get(i).getDiscriminatorValue());
+            }
+        }
+        final AnnotationSpec jsonSubTypesAnnotation = AnnotationSpec.builder(JsonSubTypes.class)
+                .addMember("value", annotationBodyBuilder.build())
+                .build();
+
+        return Stream.of(jsonSubTypesAnnotation, jsonTypeInfoAnnotation);
+    }
+
+    protected Stream<AnnotationSpec> getAdditionalTypeAnnotations(final ObjectType object) {
```

Comment:
Here we should rename the parameter to `objectType` too.

Reply:
nice catch 👍